presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java

```diff
@@ -65,48 +78,221 @@ public CachingCostProvider(CostCalculator costCalculator, StatsProvider statsPro
+    }
+
+    private static class EstimationContext
+    {
+        private final double localMaxMemory;
+        private final List<PlanCostEstimate> sourcesCosts;
+
+        public EstimationContext(double localMaxMemory, List<PlanCostEstimate> sourcesCosts)
+        {
+            this.localMaxMemory = localMaxMemory;
+            this.sourcesCosts = requireNonNull(sourcesCosts, "sourcesCosts is null");
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("localMaxMemory", localMaxMemory)
+                    .add("sourcesCosts", sourcesCosts)
+                    .toString();
+        }
+    }
+
+    private static class MemoryEstimate
+    {
+        private final double maxMemoryBeforeOutputting;
+        private final double maxMemoryWhenOutputting;
+
+        public MemoryEstimate(double maxMemoryBeforeOutputting, double maxMemoryWhenOutputting)
+        {
+            this.maxMemoryBeforeOutputting = maxMemoryBeforeOutputting;
+            this.maxMemoryWhenOutputting = maxMemoryWhenOutputting;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("maxMemoryBeforeOutputting", maxMemoryBeforeOutputting)
+                    .add("maxMemoryWhenOutputting", maxMemoryWhenOutputting)
+                    .toString();
+        }
+    }
+
+    private static class PeakMemoryEstimator
+            extends PlanVisitor<MemoryEstimate, EstimationContext>
+    {
+        @Override
+        protected MemoryEstimate visitPlan(PlanNode node, EstimationContext context)
+        {
+            if (context.sourcesCosts.isEmpty()) {
+                return new MemoryEstimate(context.localMaxMemory, context.localMaxMemory);
+            }
+
+            // A PlanNode that is estimated to use memory should have an explicit rule here.
+            verify(isNaN(context.localMaxMemory) || context.localMaxMemory == 0, "Unexpected local memory usage estimation for %s: %s", node, context.localMaxMemory);
+            // A PlanNode that has more than 1 source should have an explicit rule here.
+            double sourceMaxMemoryWhenOutputting = getOnlySourceMaxMemoryWhenOutputting(node, context);
+            return new MemoryEstimate(sourceMaxMemoryWhenOutputting, sourceMaxMemoryWhenOutputting);
+        }
+
+        @Override
+        public MemoryEstimate visitGroupReference(GroupReference node, EstimationContext context)
+        {
+            throw new IllegalStateException("Unexpected GroupReference");
+        }
+
+        @Override
+        public MemoryEstimate visitExchange(ExchangeNode node, EstimationContext context)
+        {
+            PlanCostEstimate sourcesCost = context.sourcesCosts.stream()
+                    .reduce(PlanCostEstimate.zero(), CachingCostProvider::addSiblingsCost);
+
+            return new MemoryEstimate(
```

Comment:
why are these the costs?

Reply:
Please elaborate?

Reply:
I was just asking if you could explain where the numbers come from.